packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,332 @@
+    constants.put("totalMemory", memInfo.totalMem);
+
+    DeviceType mDeviceType = getDeviceType(mContext);
+    constants.put("deviceType", mDeviceType.getValue());
+    constants.put("isTablet", mDeviceType.getValue().equals("Tablet"));
+
+    return constants;
+  }
+
+  private String getSystemName() {
+    String systemName = "";
+    if (android.os.Build.VERSION.SDK_INT < 23) {
+      systemName = "Android";
+    } else {
+      systemName = Build.VERSION.BASE_OS;
+      if (systemName.length() == 0) {
+        systemName = "Android";
+      }
+    }
+    return systemName;
+  }
+
+  private static DeviceType getDeviceType(Context context) {
+    // Detect TVs via ui mode (Android TVs) or system features (Fire TV).
+    if (context.getApplicationContext().getPackageManager().hasSystemFeature("amazon.hardware.fire_tv")) {
+      return DeviceType.TV;
+    }
+
+    UiModeManager uiManager = (UiModeManager) context.getSystemService(Context.UI_MODE_SERVICE);
+    if (uiManager != null && uiManager.getCurrentModeType() == Configuration.UI_MODE_TYPE_TELEVISION) {
+      return DeviceType.TV;
+    }
+
+    // Find the current window manager, if none is found we can't measure the device physical size.
+    WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
+    if (windowManager == null) {
+      return DeviceType.UNKNOWN;
+    }
+
+    // Get display metrics to see if we can differentiate handsets and tablets.
+    // NOTE: for API level 16 the metrics will exclude window decor.
+    DisplayMetrics metrics = new DisplayMetrics();
+    windowManager.getDefaultDisplay().getMetrics(metrics);
+
+    // Calculate physical size.
+    double widthInches = metrics.widthPixels / (double) metrics.xdpi;
+    double heightInches = metrics.heightPixels / (double) metrics.ydpi;
+    double diagonalSizeInches = Math.sqrt(Math.pow(widthInches, 2) + Math.pow(heightInches, 2));
+
+    if (diagonalSizeInches >= 3.0 && diagonalSizeInches <= 6.9) {
+      // Devices in a sane range for phones are considered to be Handsets.
+      return DeviceType.HANDSET;
+    } else if (diagonalSizeInches > 6.9 && diagonalSizeInches <= 18.0) {
+      // Devices larger than handset and in a sane range for tablets are tablets.
+      return DeviceType.TABLET;
+    } else {
+      // Otherwise, we don't know what device type we're on/
+      return DeviceType.UNKNOWN;
+    }
+  }
+
+  @ExpoMethod
+  public void getSerialNumberAsync(final Promise promise) {
+
+    int permissionCheck = ContextCompat.checkSelfPermission(mContext, Manifest.permission.READ_PHONE_STATE);
+    if (permissionCheck == PackageManager.PERMISSION_GRANTED) {
+      try {
+        if (android.os.Build.VERSION.SDK_INT < 26) {
+          promise.resolve(Build.SERIAL);
+        } else {
+          promise.resolve(Build.getSerial());
+        }
+      } catch (SecurityException e) {
+        Log.e(TAG, e.getMessage());
+        promise.reject(e);
```

Comment:
add error codes to all of these promise rejections (`ERR_DEVICE_(something about the method)`)